util/genrest/gomodel/pathtemplateparser.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	pt, err = parser.parseSegments()
+	pt = append(PathTemplate{slashSegment}, pt...)
+	if err != nil {
+		return pt, err
+	}
+
+	if parser.source.ConsumeIf(':') {
+		verb, err := parser.parseLiteral()
+		if err != nil {
+			return pt, fmt.Errorf("could not parse verb")
+		}
+
+		pt = append(pt, &Segment{Kind: Literal, Value: ":"}, verb)
+	}
+
+	if parser.source.InRange() {
+		return pt, fmt.Errorf("unexpected character")
+	}
+
+	return pt, nil
+
+}
+
+// parseSegments parses a sequence of slash-delimited segments.
+func (parser *Parser) parseSegments() (PathTemplate, error) {
+	pt := PathTemplate{}
+	proceed := true
+
+	for proceed {
+		if parser.haveLastSegment {
+			return pt, fmt.Errorf("already encountered last segment")
+		}
+		segment, err := parser.parseOneSegment()
+		pt = append(pt, segment)
+		if err != nil {
+			return pt, err
+		}
+
+		if proceed = parser.source.ConsumeIf('/'); proceed {
+			pt = append(pt, slashSegment)
+		}
+	}
+	return pt, nil
+}
+
+// segmentParser is a function type that parses a specific type of segment. It returns both nil
+// error and nil Segment if the parser does not apply to the next stream of characters in the
+// source. It returns a non-nil Segment if the characters from the point at which called matched the
+// segment type.
+type segmentParser func() (*Segment, error)
+
+// parseOneSegment parses exactly one segment of the recognized types.
+func (parser *Parser) parseOneSegment() (*Segment, error) {
+	orderedParsers := []segmentParser{parser.parseVariable, parser.parseLiteral, parser.parseMultipleValue, parser.parseSingleValue}
+	for _, parse := range orderedParsers {
+		seg, err := parse()
+		if err != nil || seg != nil {
+			return seg, err
+		}
+	}
+	return nil, fmt.Errorf("could not parse path segment")
+}
+
+// parseSingleValue parses a segment with `Kind==SingleValue` (i.e. a single-segment placeholder), returning nil if not possible.
+func (parser *Parser) parseSingleValue() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reSingleValue, `^\*`)
+	return parser.parseToSegment(re, SingleValue)
+}
+
+// parseMultipleValue parses a segment with `Kind==MultipleValue` (i.e. a multiple-segment placeholder), returning nil if not possible.
+func (parser *Parser) parseMultipleValue() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reMultipleValue, `^\*\*`)
+	seg, err := parser.parseToSegment(re, MultipleValue)
+	if seg != nil {
+		parser.haveLastSegment = true
+	}
+	return seg, err
+}
+
+// parseLiteral parses a segment with `Kind==Literal`, returning nil if not possible.
+func (parser *Parser) parseLiteral() (*Segment, error) {
+	re := parser.GetRegexp(&parser.reLiteral, "^([a-zA-Z0-9_%]*)")
+	return parser.parseToSegment(re, Literal)
+}
+
+// parseToSegment is a helper function that creates a Segment of the specified kind if the next
+// characters in the parse stream match the expression re.
+func (parser *Parser) parseToSegment(re *regexp.Regexp, kind SegmentKind) (*Segment, error) {
```

Comment:
It's standard to return `error`, but you're right, since it's never non-`nil` right now. I can omit it. If this ever starts being able to generate an error, I'll add it back.